### PR TITLE
`Programming exercises`: Allow students to download tests after due date

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage)..
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Students want to be able to understand their assessment and also find and verify their mistakes after the due date. Currently Artemis only supports exporting an example solution, but since there might be multiple different solutions, the students should also be able to export the tests and execute them on their local machines.

### Description

This PRs adds an option in the settings of a programming exercise to also publish the tests together with the example solution. If this option is enabled, a second download button for the tests is added at the bottom of the student view after the deadline has passed. This download will include the tests, the example solution in the correct place so students know where they should put their solution as well as potential auxiliar repository that should be checked out.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise with a due date and publication date for the examples

1. Log in to Artemis
2. Create the exercise and enable the publication of the tests
3. Wait until the deadline passes
4. Download the tests and inspect the .zip

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseExportService.java | 62% | 85% |

### Screenshots

Student view for downloading:
Bildschirmfoto 2022-10-23 um 22 05 28
Instructor view to enable the publishing of tests:
Bildschirm­foto 2022-11-25 um 13 45 33

Instructor view of the timeline:
Bildschirm­foto 2022-11-26 um 11 12 10

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 1]
```diff
@@ -1233,31 +1233,31 @@
...
         var zipFilenameWithoutSlash = zipFilename.replaceAll("\\s", "");
 
         if (!zipFilenameWithoutSlash.endsWith(".zip")) {
             zipFilenameWithoutSlash += ".zip";
         }
 
-        Path zipFilePath = Path.of(repositoryDir, zipFilenameWithoutSlash);
-        Files.createDirectories(Path.of(repositoryDir));
-        return zipFileService.createZipFileWithFolderContent(zipFilePath, repository.getLocalPath(), contentFilter);
+        Path zipFilePath = Path.of(zipDir, zipFilenameWithoutSlash);
```
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:1258]: This is ok, TAs are allowed to access the repos of the exercise

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 2]
```diff
@@ -130,6 +130,9 @@ public class ProgrammingExercise extends Exercise {
     @Column(name = "branch", table = "programming_exercise_details")
     private String branch;
 
+    @Column(name = "release_tests_after_due_date", table = "programming_exercise_details")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:133]: release_tests_with_example_solution

**Comment by evgeny1984** [comment]
I get an internal server errors when I deploy the current version. Also, multiple checks failed.

**Review by timo07k — CHANGES_REQUESTED** [review]
Unfortunately I instantly get a server error. Tested on TS1.

image

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove

**Review by evgeny1984 — CHANGES_REQUESTED** [review]
When I click on the button **Download tests with example solution**, I get an internal error.

**Review by evgeny1984 — DISMISSED** [review]
Tested on the ts2, works as expected, no internal errors anymore. The code changes also looks good. Approved.

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 3]
```diff
@@ -299,7 +302,11 @@
                 "disallowComplaintOnAutomaticAssessmentTooltip": "Students cannot complain after receiving an automatic assessment",
                 "complaintOnAutomaticAssessmentEnabled": "Complaints are possible",
                 "complaintOnAutomaticAssessmentDisabled": "Complaints are not possible",
-                "exampleSolutionPublicationDateTooltip": "The date when the Solution Repository becomes available to download for students. Leave blank to disable."
+                "exampleSolutionPublicationDateTooltip": "The date when the Solution Repository becomes available to download for students. Leave blank to disable.",
+                "releaseTestsWithExampleSolution": "Publish tests",
+                "releaseTestsWithExampleSolutionTooltip": "Additionally to the example solution the automatic tests can be published after the deadline.",
```
Comment (frederik-kt) [inline src/main/webapp/i18n/en/programmingExercise.json:307]:
```suggestion
                "releaseTestsWithExampleSolutionTooltip": "In addition to the example solution the automatic tests can be published after the deadline.",
```

**Review by frederik-kt — CHANGES_REQUESTED** [review]

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 4]
```diff
@@ -299,7 +302,11 @@
                 "disallowComplaintOnAutomaticAssessmentTooltip": "Die Studierenden dürfen keine Beschwerde nach einer automatischen Bewertung einlegen",
                 "complaintOnAutomaticAssessmentEnabled": "Beschwerden sind möglich",
                 "complaintOnAutomaticAssessmentDisabled": "Beschwerden sind nicht möglich",
-                "exampleSolutionPublicationDateTooltip": "Das Datum, an dem das Lösungsrepository für Studenten zum Herunterladen verfügbar wird. Zum Deaktivieren leer lassen."
+                "exampleSolutionPublicationDateTooltip": "Das Datum, an dem das Lösungsrepository für Studenten zum Herunterladen verfügbar wird. Zum Deaktivieren leer lassen.",
+                "releaseTestsWithExampleSolution": "Tests veröffentlichen",
+                "releaseTestsWithExampleSolutionTooltip": "Die automatischen Tests veröffentlichen können zusätzlich zu dem Lösungsvorschlag nach der Deadline zum Download bereit gestellt werden.",
```
Comment (danielbode) [inline src/main/webapp/i18n/de/programmingExercise.json:307]:
```suggestion
                "releaseTestsWithExampleSolutionTooltip": "Die automatischen Tests können zusätzlich zu dem Lösungsvorschlag nach der Deadline zum Download bereit gestellt werden.",
```

**Review by danielbode — CHANGES_REQUESTED** [review]
Tested on ts2. Works as expected. There is just one typo in the german tooltip.

**Review by rriyaldhi — APPROVED** [review]
Tested on TS2. Works as expected.

**Review by MarkusPaulsen — APPROVED** [review]
Worked when tested on ts2.

**Comment by JohannesStoehr** [comment]
Newest commit only changed the translations based on the feedback. Reviews are still valid

**Review by danielbode — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 5]
```diff
@@ -97,9 +97,9 @@ public final class Constants {
     public static final long MAX_SUBMISSION_FILE_SIZE = 8 * 1024 * 1024; // 8 MB
 
     // Note: The values in input.constants.ts (client) need to be the same
-    public static final int MAX_SUBMISSION_TEXT_LENGTH = 30 * 1000; // 30.000 characters
+    public static final int MAX_SUBMISSION_TEXT_LENGTH = 30_000; // 30.000 characters
 
-    public static final int MAX_SUBMISSION_MODEL_LENGTH = 100 * 1000; // 100.000 characters
+    public static final int MAX_SUBMISSION_MODEL_LENGTH = 100_000; // 100.000 characters
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:102]: do we need that comment?

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts [hunk 6]
```diff
@@ -249,21 +249,23 @@ export class CourseExerciseDetailsComponent implements OnInit, OnDestroy {
...
-            }
-        } else if (newExercise.type === ExerciseType.TEXT || newExercise.type === ExerciseType.FILE_UPLOAD) {
-            const exercise = newExercise as TextExercise & FileUploadExercise;
-            if (exercise.exampleSolution) {
-                this.exampleSolution = this.artemisMarkdown.safeHtmlForMarkdown(exercise.exampleSolution);
-            }
-        } else if (newExercise.type === ExerciseType.PROGRAMMING) {
-            const exercise = newExercise as ProgrammingExercise;
-            this.isProgrammingExerciseExampleSolutionPublished = exercise.exampleSolutionPublished || false;
+        switch (newExercise.type) {
```
Comment (simon-hng) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:253]: Please no switch cases

**Review by simon-hng — DISMISSED** [review]
Left some minor remarks, otherwise looks good

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts (hunk 6 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:253]: Is there any reason for your dislike of switch cases?

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java (hunk 5 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/config/Constants.java:102]: It specifies that it is about chars. I don't see any harm from this comment either

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts (hunk 6 above)
Comment (simon-hng) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:253]: > It’s hard to make a small switch statement.6 Even a switch statement with only two cases is
> larger than I’d like a single block or function to be. It’s also hard to make a switch state-
> ment that does one thing. By their nature, switch statements always do N things.
Robert C. Martin: Clean Code page 37

Comment (JohannesStoehr) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:253]: This is already an improvement over the if-else if... construct in my eyes and it's not the goal to refactor the complete component.
So I'd leave it like that

Comment (simon-hng) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:253]: But it's more of a personal opinion and up to you if you want to keep it or not

**Review by rriyaldhi — APPROVED** [review]
Reapprove

**Review by simon-hng — APPROVED** [review]
reapprove code

**Review by danielbode — APPROVED** [review]

**Comment by simon-hng** [comment]
image
add a margin above "Example solution" so that the grouping is more clear

**Review by TheZoker — APPROVED** [review]
LGTM 👍

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:1258]: Just to make sure: Is this enforced in the Resource or is this not actually enforced and could be changed by modifying the rest call?

**Review by TheZoker — APPROVED** [review]

**Review by danielbode — APPROVED** [review]

**Review by rriyaldhi — APPROVED** [review]

**Review by pellumbbaboci — APPROVED** [review]
Tested on ts1 during testing session, LGTM

**Review by minrows — APPROVED** [review]
Tested on ts1 during the testing session. LGTM 👍

**Review by evgeny1984 — APPROVED** [review]
Re approve after some small changes, tested on ts1.

**Pull request merged**